### ISSUE TITLE
TorchScript bug: torch.nn.transformer gives inconsistent results after conversion to TorchScript

### ISSUE THREAD

**Issue by ChenchaoZhao** [pytorch/pytorch#87368]
### 🐛 Describe the bug

torch.nn.transformer gives inconsistent results after conversion to TorchScript

```python
m = nn.TransformerEncoderLayer(3, 1, 8, batch_first=True)  # or batch_first=False 
s = torch.jit.script(m)

input_1d = torch.randn(1, 5, 3)
assert torch.allclose(m(input_1d), s(input_1d))
```

### Versions

PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.102)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.12 (main, Apr  5 2022, 01:52:34)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.6-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.12.1
[pip3] torchliter==0.3.3
[pip3] torchtext==0.13.1
[pip3] torchvision==0.13.1
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] torch                     1.12.1                   pypi_0    pypi
[conda] torchliter                0.3.3                    pypi_0    pypi
[conda] torchtext                 0.13.1                   pypi_0    pypi
[conda] torchvision               0.13.1                   pypi_0    pypi

**Comment by ingambe** [comment]
Same issue on Linux using GPU
PyTorch version: 1.12.1
Cuda 11.3
Python 3.9.13

**Comment by ChenchaoZhao** [comment]
There are dropout layers in transformer modules.
If set module and scripted module to eval, then the outputs are consistent.
